src/web100srv.c

```diff
@@ -1315,15 +1110,66 @@ int run_test(tcp_stat_agent *agent, Connection *ctl, TestOptions *testopt,
              vars.Timeouts, vars.Sndbuf, vars.MaxRwinRcvd, vars.CurrentCwnd,
              mismatch, bad_cable, totalcnt, refresh);
   }
-  if (ctl->ssl != NULL) {
-    SSL_shutdown(ctl->ssl);
-  } else {
-    shutdown(ctl->socket, SHUT_WR);
-  }
-  /* shutdown(ctlsockfd, SHUT_RDWR); */
+  close_connection(ctl);
+  shutdown(ctl->socket, SHUT_WR);
   return (0);
 }
 
+/** Read a single message sent from the parent server process to the child
+ * process, and perform and necessary communication that message necessitates.
+ * @param parent_pipe where to read the message from
+ * @param ctl the Connection to use for the network communication
+ * @param testopt the parameters of the test
+ * @param t_opt the tests the client supports
+ * @return the message value on success, a negative error code on failure.
+ */
+int read_from_parent(int parent_pipe, Connection *ctl, TestOptions *testopt,
+                     int t_opts) {
+  ParentChildSignal parent_message;
+  int retcode;
+  char buff[16];  // Enough of a buffer to hold an int.
+  int msg_len, msg_type;
+  retcode = read(parent_pipe, &parent_message, sizeof(ParentChildSignal));
+  log_println(6, "Child %d got '%d' from parent", getpid(), parent_message);
+  if (retcode == -1) {
+    retcode = errno;  // Save errno in case log_println sets it.
+    log_println(1, "Child %d couldn't read from the pipe.");
+    return -retcode;
+  }
+  if (parent_message < 0) {
+    // The message received is a negative number. This should never happen.
+    log_println(1, "Child %d got a negative number (%d) from the parent.",
+                getpid(), parent_message);
+    return -EBADMSG;
+  }
```

Comment:
Can we refactor this into separate reading and processing functions? Both halves are almost completely independent and use disjoint sets of the variables and parameters (except for `parent_message`).